CORE-9451 simplify crypto service configuration handling

dickon opened the pull request:
Previously the crypto code was tightly coupled with the config service. In particular:

1. The crypto service created its own default configuration, with embedded passphrase, and then for the combined worker stored this in its database. (Everything else in Corda uses the Config schema to handle defaults). Special measures were taken to make sure the default master wrapping key passphrase was encrypted, which required knowledge of how smart config handles secrets which is no longer safe to assume since we allow different secrets implementations.
2. Different crypto service providers were supported, each with its own configuration. In order to handle this in a general way, the smart configuration for the crypto service was converted to JSON by CryptoServiceFactoryImpl, then dynamically instantiated into configuration objects particular to a specific crypto service provider, and passed in dynamically typed. This also involved care to make sure that secrets such as the passphrase could still be decrypted, which make assumptions that secrets are always encrypted locally. With other sources of secrets that assumption is no longer sound.
3. The configuration of the crypto service providers was handled by a layer of data classes allowing plain access to the classes, rather than using methods of SmartConfig. That's reasonable of itself, but is different to the rest of Corda. 
4. Some of the crypto code at lib level was configurable, which would normally have caused a dependency on the confgiuration library. However, the configuration library has a dependency on the crypto library, in order to handle cryptogrpahy on the the secret parts of the config. This prevented some parts of the crypto code from using SmartConfig.
5. There is a separate Corda crypto technical compliance kit which shares code with the production implementation.

In combination, this means that the configuratioon of crpto scheme is complex and brittle, and would get significantly more complex when extended to include support for third party secret stores. 

In this change, I've moved to mostly using SmartConfig directly, much like the rest of Corda. I've also moved the library-level generic (templated) crypto service provider, and instead we have the components/crypto/crypto-softhsm-impl provide a simplified, non-templated/non-generic crypto service provide single abstract method interface and a simple implementation which provides a simplified type for the crypto service.  We may well never need the flexbility of different config service providers, and it is expensive to maintain and would have made this change impractical.

Having components/crypto/crypto-softhsm-impl define the CryptoServiceProvider isn't ideal, but:
a) it fits there without introducing a dependency cycle with crypto-service-impl
b) we plan to shortly collapse down the support for multiple different types of HSM, and merge crypto-softhsm-impl and crypto-sevice-impl

The Crypto service at this point still has its own way of constructing default configuration to go into the database. We may get rid of this soon, and instead simply use the JSON config schema as the rest of the Corda does. For now we pass in the passphrae and salt default values.

Extend some comments and a README on how config works based on what I figured out about configuration to design and implement this change. 

Add a toString() to dump out AesKeys in Base64. To be used with care, but if you need to actually compare keys to debug this helps.

Delete KeyCredentials, its tests and methods in other classes with initialise based on it, since we no longer use it. I simply
have the createDefaultCryptoConfig take two parameters instead of needing a dataclass.

driessamyn commented on processors/member-processor/src/integrationTest/kotlin/net/corda/processor/member/MemberProcessorTestUtils.kt:
```diff
@@ -279,7 +277,7 @@ class MemberProcessorTestUtils {
         fun Publisher.publishMembershipConf(membershipConfig: SmartConfig) =
             publishConf(ConfigKeys.MEMBERSHIP_CONFIG, membershipConfig.root().render())
-        fun Publisher.publishDefaultCryptoConf(cryptoConfig: SmartConfig) =
+        fun Publisher.publishDefaultCryptoConf(cryptoConfig: Config) =
```
I don't think this change is necessary.

dickon replied:
Yes, it was left over from an earlier direction I took and then largley reverted. Fixed.

driessamyn reviewed (commented):
Looks good to me, just one minor comment, but will wait for @kyriathar 's review.

kyriathar commented on applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt:
```diff
@@ -98,6 +98,7 @@ class CryptoWorkerParams {
     @CommandLine.Option(names = ["-d", "--database-params"], description = ["Database parameters for the worker."])
     var databaseParams = emptyMap<String, String>()
 
+    // TODO - delete me
     @CommandLine.Option(names = ["--hsm-id"], description = ["HSM ID which is handled by this worker instance."])
     var hsmId = ""
```
I think this refers to the hard HSMs which we will not support so I agree it should be removed. Could we please raise a JIRA ticket and reference it from here if not intended to remove in this PR?

dickon replied:
I don't think we need the overhead of a JIRA ticket for each tech debt TODO. I'll clarify the comment though, it's too vague.

dickon replied:
Jira Link added

kyriathar commented on applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt:
```diff
@@ -52,17 +52,29 @@ class WorkerHelpers {
             return params
         }
 
+        data class BootstrapConfigResults(
+            val bootstrapConfig: SmartConfig,
+            val smartConfigFactory: SmartConfigFactory
+        )
```
I think `SmartConfigFactory` is already available through `SmartConfig.factory` so maybe we don't need this data class.

driessamyn replied:
Good spot. We don't need this I think.

dickon replied:
Yes, nice, done.

kyriathar commented on applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt:
```diff
@@ -52,17 +52,29 @@ class WorkerHelpers {
          * Uses [smartConfigFactory] to create a `SmartConfig` containing the instance ID, topic prefix, additional
-         * params in the [defaultParams], and any [extraParams].
+         * params in the [defaultParams], and any [extraParams]. Check that the configuration matches the schema,
+         * but do not fill in defaults.
```
nitpicky: I think there are mentioned some implementation details here. Could we simplify to say: 

"Creates bootstrap configuration out of command line configuration and gets validated against schema: `corda.boot.json`

It is then passed to workers' processors and also to `ConfigurationReadService` (see `ConfigurationReadService.bootstrapConfig`)."

Maybe also briefly say what it contains?

dickon replied:
The comment does address is implementation details. I don't see any problem with having such material in this comment, nor do I see anywhere else to put such a comment.  If this was split out to an interface and a subclass that implements the interface then you'd have a point. I did take the rest of your comment.

kyriathar commented on applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt:
```diff
@@ -87,13 +99,29 @@ class WorkerHelpers {
             validator.validate(BOOT_CONFIG, bootConfig, loadResource(BOOT_CONFIG_PATH))
-            return bootConfig
+
+            // we now know bootConfig has:
+            //
+            // 1. parameters passed directly in on the command line
+            // 2. INSTANCE_ID and TOPIC_PREFIX set indirectly via the command line
+            //
+            // Also, boot config has been validated; all the keys in it are declared in the JSON schema and are of
+            // the types specified in the schema
+            //
+            // However, bootConfig does not:
+            //  - have database configuration records applied
+            //  - have unspecified fields filled in with defaults from the schema.
+            //
+            // This is handled later, via the lifecycle coordinator and the domino logic,
+            // based on the requirements graph.
+            //
```
I think maybe remove this comment? First part should be mentioned on kDoc. Then the defaulting I think you mean happens later. Here the defaulting is whatever comes in from for `DefaultWorkerParams` and workers specific defaulted fields.

dickon replied:
I updated the comment to clarify a little.

kyriathar commented on applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt:
```diff
@@ -87,13 +99,29 @@ class WorkerHelpers {
                 ConfigFactory.parseMap(defaultParams.secretsParams.mapKeys { (key, _) -> "${ConfigKeys.SECRETS_CONFIG}.${key.trim()}" })
 
-            val bootConfig = SmartConfigFactory
+            val smartConfigFactory = SmartConfigFactory
```
We could remove this variable and inline with `bootConfig`.

dickon replied:
This changed a bit as a result of using SmartConfig.factory; I think it is okay now.

kyriathar commented on components/configuration/README.md:
```diff
@@ -0,0 +1,93 @@
+
+Corda is dynamic enough that the configuration needs to be dynamic as well. For instance, in some situations,
+Corda can create new database credentials which we store in the configuration system. So as well as external
+soruces of configuration, we also write configuration records to a to Kafka so that they can be passed around multiple
```
```suggestion
sources of configuration, we also write configuration records to Kafka so that they can be passed around multiple
```

dickon replied:
done, thanks

kyriathar commented on components/configuration/README.md:
```diff
@@ -0,0 +1,93 @@
+1. A worker starts as a top level JVM command line program, via OSGi, and gets its command line arguments.
+   This might be a single-purpose worker such as the RPC worker, or, in the non-production local development mode  
+   might be the combined worker.
+2. `WorkerHelpers.getBootstrapConfig` is called to put together a validated smart config object. This
+   bootstrap configuration will typically be incomplete since it will only contain keys specified on the command line,
+   rather than all keys specified in the JSON schema of the configuration in the `corda-api.git/data/config-schema`,
+   but what is in it is guaranteed to match the schema, or else this step will fail.
```
I believe when you say "all keys specified" you mean dynamic config coming in from RPC. So different than bootstrap config. I think this section might be a bit misleading.

dickon replied:
I don't think it's misleading, and I'm not sure what you are getting at.

kyriathar commented on components/configuration/README.md:
```diff
@@ -0,0 +1,93 @@
+Corda makes use of the [typesafe config](https://github.com/lightbend/config) configuration library. We extend
```
I think let's remove this README file from here and lets make it a confluence wiki page. Then we can review there and potentially add it back here. I think, it is currently very detailed making it a bit hard to follow. Also might become obsolete if any of the steps changes.

driessamyn replied:
I think some of this is already captured in the github wiki, so may you're right, maybe it should go there.

dickon replied:
It's not clear that there's any reason to keep this internal. I agree it is tightly coupled to the code, which was one reason I put it here. I'm not sure how the github wiki handles branching, but usually wikis aren't good at that,  so the reason we still have significant detail in the READMEs is to give us the opportunity and encouragement to keep it in sync with the code.  Github wikis are good for point-in-time text. 

I don't think any of the detail here is irrelevant (and feel free to quibble on particular clauses in the text if you disagree) if you want to understand how configuration works, and configuration is important.  Simplification would be welcome, although I don't see much opportunity given our requirements.

driessamyn replied:
The Github wiki isn't internal.

I think the question is really about how many places we have developer documentation in.

I agree it doesn't feel right to be here as it is inconsistent with other documentation.
I feel like it's too  much detail as well, which will go out of date quickly, but I appreciate that's subjective.

kyriathar commented on components/configuration/configuration-read-service/src/main/kotlin/net/corda/configuration/read/ConfigurationReadService.kt:
```diff
@@ -67,12 +67,14 @@ interface ConfigurationReadService : Lifecycle {
-     * @throws ConfigurationException If the bootstrap configuration is provided after a connection has been
-     *                                established, or if the configuration does not allow access.
+     *  @param config The bootstrap configuration to connect to the message bus.
+     *  @throws ConfigurationException If the bootstrap configuration is provided after a connection has been
```
I would revert these kDocs.

dickon replied:
:+1: done

kyriathar commented on applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt:
```diff
@@ -94,7 +94,16 @@ class CombinedWorker @Activate constructor(
         val cryptoConfig = PathAndConfig(BOOT_CRYPTO, createCryptoBootstrapParamsMap(params.hsmId))
-        val config = getBootstrapConfig(
+
+        // getBootstrapConfig involves setting up a smart config factory, in order to provide access to
+        // the boot configuration. So, to prevent having to set up smart config factory in several places,
+        // we have getBootstrapConfig return the smart config factory as well as a smart config instance representing
+        // the boot config when we get hold of the boot config.
+        //
+        // It may be worth separating out the smart config factory setup so that we can pass it in to both
+        // getBootstrapConfig and the PostgresDbSetup below.
```
Should we remove this comment?

dickon replied:
Why?

driessamyn replied:
Is it still valid as the factory is available as a property on the SmartConfig?

kyriathar commented on libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt:
```diff
@@ -160,7 +156,7 @@ private const val CRYPTO_CONNECTION_FACTORY_OBJ = "cryptoConnectionFactory"
 private const val HSM_SERVICE_OBJ = "hsmService"
 private const val HSM_MAP = "hsmMap"
-private const val HSM_MAP_ITEM_OBJ = "hsmMap.%s"
+const val HSM_MAP_ITEM_OBJ = "hsmMap.%s"
```
I think revert back to private? Seems to only getting used in here.

dickon replied:
yeah done

kyriathar commented on tools/plugins/initial-config/src/main/kotlin/net/corda/cli/plugin/initialconfig/CryptoConfigSubcommand.kt:
```diff
@@ -59,25 +60,27 @@ class CryptoConfigSubcommand : Runnable {
+            ${EncryptionSecretsServiceFactory.SECRET_SALT_KEY}=${salt}
+        """.trimIndent()
+            ),
+            listOf(EncryptionSecretsServiceFactory())
```
My understanding is, in this cli tool we statically use `EncryptionSecretsServiceFactory`. But Corda could be using some different one? In which case it will not be able to decrypt these secrets?

dickon replied:
Yeah, comment and ticket added.

kyriathar commented on libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/aes/KeyCredentials.kt:
```diff
@@ -1,18 +0,0 @@
-package net.corda.crypto.core.aes
```
Removing this class has increased the changes footprint of this PR I think.

dickon replied:
Right, but it's also dead code, since it was only used in this one place, so to not do so would have increased tech debt.

kyriathar reviewed (commented):
Some comments, still need to look at dynamic config that we removed.

driessamyn commented on .run/Combined Worker Local.run.xml:
```diff
@@ -2,11 +2,12 @@
     <option name="JAR_PATH" value="$PROJECT_DIR$/applications/workers/release/combined-worker/build/bin/corda-combined-worker-5.0.0.0-SNAPSHOT.jar" />
     <option name="VM_PARAMETERS" value="-Dco.paralleluniverse.fibers.verifyInstrumentation=true" />
-    <option name="PROGRAM_PARAMETERS" value="--instance-id=0 -mbus.busType=DATABASE -spassphrase=password -ssalt=salt -spassphrase=password -ssalt=salt -ddatabase.user=user -ddatabase.pass=password -ddatabase.jdbc.url=jdbc:postgresql://localhost:5432/cordacluster -ddatabase.jdbc.directory=$ProjectFileDir$/applications/workers/release/combined-worker/drivers" />
+    <option name="PROGRAM_PARAMETERS" value="--instance-id=0 -mbus.busType=DATABASE -ddatabase.user=user -ddatabase.pass=password -ddatabase.jdbc.url=jdbc:postgresql://localhost:5432/cordacluster -ddatabase.jdbc.directory=$ProjectFileDir$/applications/workers/release/combined-worker/drivers -p 7004 -stype=hashicorp-vault -surl=http://127.0.0.1:8200 -stoken=myroot -daddon.directory=$PROJECT_DIR$/. ...
```
This should not be committed.

dickon replied:
Yes, that was probabl a mistake in git rebasing, glad you spotted it.

dickon replied:
reverted

ronanbrowne reviewed (dismissed):
LGTM - build

dickon committed:
CORE-9451: rework crypto config

Previously the crypto code was tightly coupled with the config service. In particular:

The crypto service created its own default configuration, with embedded passphrase, and then for the combined worker stored this in its database. (Everything else in Corda uses the Config schema to handle defaults). Special measures were taken to make sure the default master wrapping key passphrase was encrypted, which required knowledge of how smart config handles secrets which is no longer safe to assume since we allow different secrets implementations.
Different crypto service providers were supported, each with its own configuration. In order to handle this in a general way, the smart configuration for the crypto service was converted to JSON by CryptoServiceFactoryImpl, then dynamically instantiated into configuration objects particular to a specific crypto service provider, and passed in dynamically typed. This also involved care to make sure that secrets such as the passphrase could still be decrypted, which make assumptions that secrets are always encrypted locally. With other sources of secrets that assumption is no longer sound.
The configuration of the crypto service providers was handled by a layer of data classes allowing plain access to the classes, rather than using methods of SmartConfig. That's reasonable of itself, but is different to the rest of Corda.
Some of the crypto code at lib level was configurable, which would normally have caused a dependency on the confgiuration library. However, the configuration library has a dependency on the crypto library, in order to handle cryptogrpahy on the the secret parts of the config. This prevented some parts of the crypto code from using SmartConfig.
There is a separate Corda crypto technical compliance kit which shares code with the production implementation.
In combination, this means that the configuratioon of crpto scheme is complex and brittle, and would get significantly more complex when extended to include support for third party secret stores.

In this change, I've moved to mostly using SmartConfig directly, much like the rest of Corda. I've also moved the library-level generic (templated) crypto service provider, and instead we have the components/crypto/crypto-softhsm-impl provide a simplified, non-templated/non-generic crypto service provide single abstract method interface and a simple implementation which provides a simplified type for the crypto service. We may well never need the flexbility of different config service providers, and it is expensive to maintain and would have made this change impractical.

Having components/crypto/crypto-softhsm-impl define the CryptoServiceProvider isn't ideal, but:
a) it fits there without introducing a dependency cycle with crypto-service-impl
b) we plan to shortly collapse down the support for multiple different types of HSM, and merge crypto-softhsm-impl and crypto-sevice-impl

The Crypto service at this point still has its own way of constructing default configuration to go into the database. We may get rid of this soon, and instead simply use the JSON config schema as the rest of the Corda does. For now we pass in the passphrae and salt default values.

Extend some comments and a README on how config works based on what I figured out about configuration to design and implement this change.

Add a toString() to dump out AesKeys in Base64. To be used with care, but if you need to actually compare keys to debug this helps.

Delete KeyCredentials, its tests and methods in other classes with initialise based on it, since we no longer use it. I simply
have the createDefaultCryptoConfig take two parameters instead of needing a dataclass.

driessamyn reviewed (approved):
LGTM

mag1c-48 reviewed (approved):
lgtm - build
